### PR TITLE
Crash in fetch_interface if the upgraded package modifies the view

### PR DESCRIPTION
**Pull request by paulbrauner-da**
Addresses the fetchInterface part of UP-1 in https://docs.google.com/spreadsheets/d/1_0ER--x-YJzoYSoy7VFs6emzn4ZYjBc0kte4MLosuxs/edit?gid=0#gid=0

### CODE REVIEW

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -1139,14 +1141,130 @@ private[lf] object SBuiltin {
...
+                          viewInterface(machine, interfaceId, dstTmplId, dstArg) { dstView =>
+                            executeExpression(machine, sourceView) { sourceViewValue =>
+                              // If the destination and source templates are the same, we skip the computation
+                              // of the destination template's view.
+                              if (dstTmplId == sourceTplId)
+                                cacheContractAndInsertNode(machine, coid, dstTmplId, dstArg)
+                              else
+                                executeExpression(machine, dstView) { dstViewValue =>
+                                  if (sourceViewValue != dstViewValue) {
+                                    crash("view mistmatch")
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1198]: We need a proper error.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 2]
```diff
@@ -1139,14 +1141,130 @@ private[lf] object SBuiltin {
...
+  final case class SBFetchInterface(interfaceId: TypeConName) extends UpdateBuiltin(1) {
+    override protected def executeUpdate(
+        args: util.ArrayList[SValue],
+        machine: UpdateMachine,
+    ): Control[Question.Update] = {
+      val coid = getSContractId(args, 0)
+      fetchAny(machine, None, coid, SValue.SValue.None) { (maybePkgName, sourceContract) =>
+        maybePkgName match {
+          case None =>
+            crash(s"unexpected contract instance without packageName")
```
Comment (remyhaemmerle-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1174]: crashes should be limited to unexpected error. 
What it is the case of this one ?

Comment (paulbrauner-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1174]: As discussed, this comment was about 
```
 crash(s"failed to convert contract $coid to type $dstTmplId")
 ```
below, which I have replaced with a WronglyTyped error.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (paulbrauner-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: Used a Dev error for now, will discuss with @tudor-da how to introduce a new ledger API error.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -799,37 +799,40 @@ private[lf] final class Compiler(
...
     topLevelFunction2(t.FetchInterfaceDefRef(ifaceId)) { (cidPos, _, env) =>
-      let(
-        env,
-        (if (soft) SBSoftFetchInterface else SBFetchAny(None))(
-          env.toSEVar(cidPos),
-          s.SEValue.None,
-        ),
-      ) { (payloadPos, env) =>
+      if (soft) {
+        SBFetchInterface(ifaceId)(env.toSEVar(cidPos))
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:803]: I'll rename it to something that contains the word "Soft" in the PR that updates exercise by interface, and I'll share the code between the two.

**Comment by paulbrauner-da** [comment]
Superseded by https://github.com/digital-asset/daml/pull/19857

**Pull request closed without merging**